### PR TITLE
Add complete_info()

### PR DESCRIPTION
**Pull request by Shougo**
I have added `complete_mode()` API.
It returns current ctrl_x mode as string.

It is useful for auto completion plugin.

### CODE REVIEW

File: src/edit.c [hunk 1]
```diff
@@ -3003,6 +3003,63 @@ set_completion(colnr_T startcol, list_T *list)
...
 
+// complete_mode() implementation.
+char *ins_compl_mode(void)
+{
+    char *mode = NULL;
+    switch (ctrl_x_mode)
+    {
+	case 0:
+	    // Keyword completion
+	    mode = strdup("keyword");
```
Comment (dpelle) [inline src/edit.c]: Vim does not use strdup() which is not portable. It uses vim_strsave() instead.

File: src/testdir/test_popup.vim [hunk 2]
```diff
@@ -896,4 +896,48 @@ func Test_menu_only_exists_in_terminal()
...
+        \ ["\<C-X>\<C-O>", 'omni'],
+        \ ["\<C-X>s", 'spell'],
+        \]
+    call feedkeys("i" . key . "\<f5>\<ESC>", 'tx')
+    call assert_equal(expected, getline('.'))
+    %d
+  endfor
+  call delete('dummy.txt')
+
+  function! s:complTestEval() abort
```
Comment (dpelle) [inline src/testdir/test_popup.vim]: Vim tests no longer uses the exclamation mark in "function!" but uses "func" … "endfunc".

File: src/testdir/test_popup.vim [hunk 3]
```diff
@@ -896,4 +896,48 @@ func Test_menu_only_exists_in_terminal()
...
+    call complete(1, ['source', 'soundfold'])
+    return ''
+  endfunction
+  inoremap <F6>  <c-r>=s:complTestEval()<CR>
```
Comment (dpelle) [inline src/testdir/test_popup.vim]: The test has a side-effect: it leaks this map. We should do iunmap <F6> before exiting the test.

File: src/testdir/test_popup.vim [hunk 4]
```diff
@@ -896,4 +896,48 @@ func Test_menu_only_exists_in_terminal()
   endtry
 endfunc
 
+func Test_popup_complete_mode()
+  new
+  inoremap <f5> <c-r>=complete_mode()<cr>
```
Comment (dpelle) [inline src/testdir/test_popup.vim]: The test has a side-effect: it leaks this map. We should do iunmap <F5> before exiting the test.

File: src/testdir/test_popup.vim [hunk 5]
```diff
@@ -896,4 +896,48 @@ func Test_menu_only_exists_in_terminal()
...
+  call writefile([
+        \ '!rm	src/testdir/amiga.vim	/^cmap !rm !Delete all$/;"	m',
+        \], 'dummy.txt')
```
Comment (dpelle) [inline src/testdir/test_popup.vim]: In general, temporary files created by Vim tests start with X presumably because it makes it easier to spot temporary files which were not removed.

File: runtime/doc/eval.txt [hunk 6]
```diff
@@ -3522,6 +3523,29 @@ complete_check()				*complete_check()*
 		Only to be used by the function specified with the
 		'completefunc' option.
 
+							*complete_mode()*
+complete_mode()
+		Return a string that indicates the current completion mode.
+		Note: It does not check popup menu is visible.
```
Comment (dpelle) [inline runtime/doc/eval.txt]: It does not check  *whether* popup menu is visible

**Comment by Shougo** [comment]
I have fixed for reviews.

**Comment by mattn** [comment]
Totally, char_u* should be used instead of char* ?

**Comment by Shougo** [comment]
OK.  I have changed.

File: src/testdir/test_popup.vim (hunk 4 above)
Comment (mattn) [reply src/testdir/test_popup.vim]: Or put `<buffer>`.

Comment (Shougo) [reply src/testdir/test_popup.vim]: Oh, it seems better.

**Comment by brammool** [comment]
> I have added `complete_mode()` API.
> It returns current ctrl_x mode as string.
> 
> It is useful for auto completion plugin.

I wonder if other completion state would be useful.  I rather have one
function to get the whole state, than a bunch of functions for several
pieces of state.

Unfortunately, the completion state is very complex, it's not easy to
even explain what each state variable means.  I've been meaning to
refactor this, but it's a complex thing.


About the implementation: Instead of using a big switch, it would be
much more compact to use a table with the completion types and string.
Something like:

	struct {
	  int mode;
	  char *name;
	} mode_table = {
	 {0, "keyword"},
	 {CTRL_X_FILES, "files"},
	};


+// complete_mode() implementation.
+char_u *ins_compl_mode(void)
+{

This should use Vim style function header.


-- 
'I generally avoid temptation unless I can't resist it."
		-- Mae West

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by Shougo** [comment]
Thank you for the review.
I have changed the implementation to use table.

**Comment by brammool** [comment]
> Thank you for the review.
> I have changed the implementation to use table.

It's a bit simpler this way:

    for (i = 0; i < sizeof(table)/ sizeof(struct mode_table); i++)
        if (table[i].mode == ctrl_x_mode)
            return (char_u *)table[i].name;

    return (char_u *)"unknown";

Please use Vim code style.

-- 
MORTICIAN:    What?
CUSTOMER:     Nothing -- here's your nine pence.
DEAD PERSON:  I'm not dead!
MORTICIAN:    Here -- he says he's not dead!
CUSTOMER:     Yes, he is.
DEAD PERSON:  I'm not!
                                  The Quest for the Holy Grail (Monty Python)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by Shougo** [comment]
OK.  I have fixed it.

**Comment by h-east** [comment]
It does not seem to be reflected here.
https://groups.google.com/d/msg/vim_dev/6utE1ObSYY8/fOaLtsBUGwAJ

**Comment by Shougo** [comment]
Oh, I have not read your response.

I have included (and some changed) your patch.

File: src/edit.c [hunk 7]
```diff
@@ -3007,33 +3007,31 @@ set_completion(colnr_T startcol, list_T *list)
...
 
-    for (i = 0; i < sizeof(table)/ sizeof(struct mode_table); i++)
-        if (table[i].mode == ctrl_x_mode)
-            return (char_u *)table[i].name;
+    if (ctrl_x_mode == CTRL_X_NOT_DEFINED_YET || compl_started)
+	mode = mode_names[ctrl_x_mode & ~CTRL_X_WANT_IDENT];
 
-    return (char_u *)"unknown";
+    return (char_u *)mode;
 }
```
Comment (mattn) [inline src/edit.c]:
```c
  if (ctrl_x_mode == CTRL_X_NOT_DEFINED_YET || compl_started)
        return mode_names[ctrl_x_mode & ~CTRL_X_WANT_IDENT];
  return (char_u *)"";
```

**Comment by h-east** [comment]
It is better to overwrite by my patch as is.
Check the contents of my patch more.

I add to the document, add test cases, simplify the test file (Xdummy.txt), add code comment, and move the definition position of the function (ins_compl_mode) to the appropriate place.

Or it may be better for Bram to include my patch :)

File: src/edit.c [hunk 8]
```diff
@@ -3589,6 +3590,36 @@ ins_compl_active(void)
     return compl_started;
 }
 
+// Return Insert completion mode name string
```
Comment (h-east) [inline src/edit.c]:
```c
/*
 * Return Insert completion mode name string
 */
```

Why do you get worse?
It has already been pointed out by Bram.
See the document.

```vim
 :h style-examples
/function_name
```

File: runtime/doc/eval.txt [hunk 9]
```diff
@@ -3522,6 +3523,30 @@ complete_check()				*complete_check()*
...
+		Note: It does not check whether popup menu is visible.
+
+		   ""		Not in completion mode
+		   "keyword"	Keyword completion |i_CTRL-X_CTRL-N|
+		   "ctrl_x"	Just press |i_CTRL-X|
+		   "whole_line"	Whole lines |i_CTRL-X_CTRL-L|
+		   "files"	File names |i_CTRL-X_CTRL-F|
+		   "tags"	Tags |i_CTRL-X_CTRL-]|
+		   "path_defines"
+				Definition completion |i_CTRL-X_CTRL-D|
```
Comment (h-east) [inline runtime/doc/eval.txt]: Why did you change it? The way it does not break is easier to see.

File: runtime/doc/eval.txt [hunk 10]
```diff
@@ -3522,6 +3523,30 @@ complete_check()				*complete_check()*
...
+		   "path_patterns"
+		   		Include completion |i_CTRL-X_CTRL-I|
```
Comment (h-east) [inline runtime/doc/eval.txt]: Same above.

File: src/proto/edit.pro [hunk 11]
```diff
@@ -20,6 +20,7 @@ char_u *find_word_end(char_u *ptr);
 int ins_compl_active(void);
 int ins_compl_add_tv(typval_T *tv, int dir);
 void ins_compl_check_keys(int frequency, int in_compl_func);
+char_u *ins_compl_mode(void);
```
Comment (h-east) [inline src/proto/edit.pro]: This is after `int ins_compl_active(void);`

File: src/testdir/test_popup.vim [hunk 12]
```diff
@@ -896,4 +896,49 @@ func Test_menu_only_exists_in_terminal()
...
+  setlocal tags=Xdummy.txt
+  setlocal dictionary=Xdummy.txt
+  setlocal thesaurus=Xdummy.txt
+  setlocal omnifunc=syntaxcomplete#Complete
+  setlocal completefunc=syntaxcomplete#Complete
+  setlocal spell
+  for [key, expected] in [
+        \ ["", ''],
+        \ ["\<C-X>", 'ctrl_x'],
+        \ ["\<C-X>\<C-N>", 'keyword'],
```
Comment (h-east) [inline src/testdir/test_popup.vim:914]: Test case is missing.
`\ ["\<C-X>\<C-P>", 'keyword'],`

File: src/edit.c [hunk 13]
```diff
@@ -3589,6 +3590,36 @@ ins_compl_active(void)
...
+	"cmdline",
+	"function",
+	"omni",
+	"spell",
+	"unknown",	    // CTRL_X_LOCAL_MSG only used in "ctrl_x_msgs"
+	"eval",
+    };
+
+    if (ctrl_x_mode == CTRL_X_NOT_DEFINED_YET || compl_started)
+	return mode_names[ctrl_x_mode & ~CTRL_X_WANT_IDENT];
```
Comment (h-east) [inline src/edit.c]: Need a cast `(char_u *)`.

**Review by h-east — COMMENTED** [review]
I think your arrangement is subtle.
So, did you decide that this is unnecessary?

```patch
diff --git a/runtime/doc/insert.txt b/runtime/doc/insert.txt
index 0431043c2..2bb357ef4 100644
--- a/runtime/doc/insert.txt
+++ b/runtime/doc/insert.txt
@@ -642,6 +642,7 @@ and one of the CTRL-X commands.  You exit CTRL-X mode by typing a key that is
 not a valid CTRL-X mode command.  Valid keys are the CTRL-X command itself,
 CTRL-N (next), and CTRL-P (previous).
 
+To get the current completion mode, |complete_mode()| can be used.
 Also see the 'infercase' option if you want to adjust the case of the match.
 
 							*complete_CTRL-E*
```

File: src/edit.c (hunk 8 above)
Comment (Shougo) [reply src/edit.c]: > It has already been pointed out by Bram.

Oh, I have not see it in the issue.
It is pointed in ML thread?

File: src/testdir/test_popup.vim (hunk 12 above)
Comment (Shougo) [reply src/testdir/test_popup.vim:914]: I have added it.

**Comment by prabirshrestha** [comment]
I would also be interested in this patch. 

I do see Bram's comment on
>I wonder if other completion state would be useful.  I rather have one
function to get the whole state, than a bunch of functions for several
pieces of state.

One option would be to have `complete_info()` that returns the following. Feel free to suggest better names.

```json
{
    "mode": "omni",
    "items": [],
    "selected": -1, // index of items
    "height": 10,
    "width": 20
}
```

For start we could only return `{"mode": "omni" }` and add others as needed. We can add docs that says to check of key before accessing it.

**Comment by Shougo** [comment]
I have renamed `complete_mode()` to `complete_info()`.

> Patch to add complete_mode().  Shougo - #3866.  Alternate patch by Hirohito
Higashi, 2019 Jan 27, included now?

@brammool I have already included Higashi's patch.

**Comment by h-east** [comment]
@Shougo 
Do NOT use my patch.
I'll Pull Request for a few days

--
Best regards,
Hirohito Higashi (h_east)

File: runtime/doc/eval.txt [hunk 14]
```diff
@@ -2267,6 +2267,7 @@ col({expr})			Number	column nr of cursor or mark
 complete({startcol}, {matches}) none	set Insert mode completion
 complete_add({expr})		Number	add completion match
 complete_check()		Number	check for key typed during completion
+complete_info()			String	get current completion information
```
Comment (prabirshrestha) [inline runtime/doc/eval.txt]: Type should be changed to `dictionary`

File: runtime/doc/eval.txt [hunk 15]
```diff
@@ -3536,6 +3537,36 @@ complete_check()				*complete_check()*
 		Only to be used by the function specified with the
 		'completefunc' option.
 
+							*complete_info()*
+complete_info()
+		Return a string that indicates the current completion
```
Comment (prabirshrestha) [inline runtime/doc/eval.txt]: dictionary instead of string

**Comment by mattn** [comment]
I think this is enough to merge.

**Comment by prabirshrestha** [comment]
Verified that this patch works for asyncomplete v2 to ignore processing autocomplete if it is open due to vim's builtin completions. :shipit:

**Comment by Shougo** [comment]
https://github.com/vim/vim/pull/4106#issuecomment-473615543

> I sent a patch that was almost completely rewritten on # 3866.
> Despite the reviews I have repeatedly made, he kept subtle.
> (and he has never said thanks-word)

OK.  Thank you for your review.
I know your patch is better.

But you have said in other place "Your patch implementation is very bad. My patch should be included instead".
So I was stubborn sorry.

I will close the PR.

Note:  He have blocked me in github.  So I cannot comment the issue.
